synir/src/data_structures/clifford_tableau.rs

```diff
@@ -191,6 +191,89 @@ impl CliffordTableau {
         }
     }
 
+    /// Composes a gadget onto a Clifford tableau if the angle is Clifford
+    /// Decomposes the Pauli gadget by performing naive decomposition into mapping to Z legs, CNOT walls and Z-rotations
+    pub fn compose_gadget(&mut self, rhs: (PauliString, Angle)) {
+        let (pauli_string, angle) = rhs;
+        let size = self.size();
+        assert_eq!(
+            size,
+            pauli_string.len(),
+            "Cannot compose Clifford tableau with PauliPolynomial of different size"
+        );
+        let pi2rotations = match angle {
+            Angle::Angle(angle) => panic!(
+                "Cannot compose Clifford tableau with non-Clifford angle: {}",
+                angle
+            ),
+            Angle::Pi4Rotations(rotations) => {
+                if rotations % 2 == 1 {
+                    panic!("Cannot compose Clifford tableau with non-Clifford angle: {} pi/4 rotations", rotations);
+                }
+                (rotations >> 1) % 4
```

Comment:
It should be an invariant that `0 <= rotations <= 7`, so `% 4` is not needed. Also, I find `/2` more explicit than a bit shift in this scenario.